docs/design-documents/2020-02-Promtail-Push-API.md

```diff
@@ -0,0 +1,124 @@
+
+1. Have a dedicated Promtail instance for receiving pushes. This also applies to
+   using the syslog target.
+2. Have a separated k8s service that always resolves to the same Promtail pod,
+   bypassing the load balancing issue.
+
+## Alternative Implementation Options
+
+The proposed feature of copying the existing Loki Push API limits how this
+feature can be implemented. The following are a list of alternative
+implementations that allow us to stray from the existing API.
+
+All alternative implementations options assume we will at least have HTTP1
+support.
+
+These options may be combined.
+
+### Option 1: JSON / Protobuf Payload
+
+A new JSON and Protobuf payload format can be designed instead of the existing
+Loki push payload. Both formats would have to be exposed to support clients that
+either can't or won't use protobuf marshalling.
+
+The primary benefit of this approach is to allow us to tweak the payload schema
+independently of Loki's existing schema, but otherwise may not be very useful
+and is essentially just code duplication.
+
+### Option 2: gRPC Service
+
+The
+[logproto.Pusher](https://github.com/grafana/loki/blob/f7ee1c753c76ef63338d53cfba782188a165144d/pkg/logproto/logproto.proto#L8-L10)
+service could be exposed through Promtail. This would enable clients stubs to be
+generated for lanuages that have gRPC support, and, for HTTP1 support, a
+[gRPC Gateway](https://github.com/grpc-ecosystem/grpc-gateway) would be embedded
+in Promtail itself.
+
+This implementation option is similar to the original proposed solution, but
+uses the gRPC gateway to handle HTTP1 traffic instead of the HTTP1 shim that
+Loki uses. There are some concerns with this approach:
+
+1. The gRPC Gateway reverse proxy will need to play nice with the existing HTTP
+   mux used in Promtail.
+2. We couldn't control the HTTP and Protobuf formats separately as Loki can.
+3. Log lines will be double-encoded thanks to the reverse proxy.
+4. A small overhead of using a reverse proxy in-process will be introduced.
+5. This breaks our normal pattern of writing our own shim functions; may add
+   some cognative overhead of having to deal with the gRPC gateway as an outlier
+   in the code.
+
+### Option 3: Plaintext Payload
+
+Prometheus' [Push Gateway API](https://github.com/prometheus/pushgateway#command-line)
+is cleverly designed and we should consider implementing our API in the same
+format: users would push to `http://promtail-url/push/label1/value1?timestamp=now`
+with a plaintext POST body. For example:
+
+```
+curl -X POST http://promtail.default/push/foo/bar/fizz/buzz -d “hello, world!”
+```
+
+This approach may be slightly faster when compared to a non-plaintext payload as
+no unmarshaling needs to be performed. This URL path and timestamp still needs
+to be parsed, but this will generally be faster than the reflection requirements
+imposed by JSON.
+
+However, note that this API limits Promtail to accepting one line at a time and
```

Comment:
I think we could support batching easily via multiline http bodies, like
```
<ts>
<log-line>
<ts>
<log-line>
...
```

Reply:
@rfratto does ^ seem reasonable to you?

Reply:
I'm not sure. I'm still partial to just having Promtail expose the same push API as Loki and leaving decisions about text-based APIs to external tooling. 

For me, I really want to be able to run this when doing local testing for something:

```
loki -log.level=debug -config.file=/my/config/file | \
  promtail-cat -lapp=loki http://localhost:1234/loki/api/v1/push
```

I think your batching idea makes sense, it just drifts a little bit from what I'm hoping to eventually be able to do with this API. (Although I suppose my `promtail-cat` tool could still exist on top of what you're suggesting here)

Reply:
Hmm, that's a good point. It seems _weird_ to expose an unbatched endpoint, but I guess it's also weird/redundant to do double buffering here and in promtail. I guess the nice part is it's easy to scale out promtails linearly across infra/streams, so the cost of unbuffered pushes is amortized over that. I think you're taking the right approach here.

Reply:
You're taking a bit bet based on preference and possibility to cat logs to promtail.

I would like a benchmark here, batching vs non batching on high throughput. We don't have a good solution for scaling and we're saying it's fine we will scale easily promtail.

Also you can now pipe data to promtail.

Reply:
Maybe I'm miscommunicating? I do want the solution we build here to support high throughput, and I do want it to be able to support batching. 

The idea of a `promtail-cat` is a completely separate project that lives outside of the Loki repo and that builds in top of an existing high-throughput solution. That separate tool isn't meant for high throughput. It's out of scope for the design here and I probably shouldn't have brought it up.